src/crypto/crypto.js

```diff
@@ -152,83 +151,141 @@ export default {
-        return [type_mpi, type_mpi, type_mpi];
       //   Algorithm-Specific Fields for DSA public keys:
       //       - MPI of DSA prime p;
       //       - MPI of DSA group order q (q is a prime divisor of p-1);
       //       - MPI of DSA group generator g;
       //       - MPI of DSA public-key value y (= g**x mod p where x  is secret).
+      //   Algorithm-Specific Fields for Elgamal public keys:
+      //       - MPI of Elgamal prime p;
+      //       - MPI of Elgamal group generator g;
+      //       - MPI of Elgamal public key value y (= g**x mod p where x  is secret).
+      case enums.publicKey.rsaEncrypt:
+      case enums.publicKey.rsaEncryptSign:
+      case enums.publicKey.rsaSign:
       case enums.publicKey.dsa:
-        return [type_mpi, type_mpi, type_mpi, type_mpi];
+      case enums.publicKey.elgamal: {
+        const params = {};
+        const paramNames = this.getPublicKeyParamNames(algo);
+        paramNames.forEach(name => {
+          const param = util.readMPI(bytes.subarray(read)); read += param.length + 2;
+          params[name] = param;
+        });
+        return { read: read, params };
+      }
       //   Algorithm-Specific Fields for ECDSA/EdDSA public keys:
       //       - OID of curve;
       //       - MPI of EC point representing public key.
-      case enums.publicKey.ecdsa:
-      case enums.publicKey.eddsa:
-        return [type_oid, type_mpi];
+      case enums.publicKey.ecdsa: {
+        const oid = new OID(); read += oid.read(bytes);
+        const Q = util.readMPI(bytes.subarray(read)); read += Q.length + 2;
+        return { read: read, params: { oid, Q } };
+      }
+      case enums.publicKey.eddsa: {
+        const oid = new OID(); read += oid.read(bytes);
+        let Q = util.readMPI(bytes.subarray(read)); read += Q.length + 2;
+        Q = util.padToLength(Q, 33);
+        return { read: read, params: { oid, Q } };
+      }
       //   Algorithm-Specific Fields for ECDH public keys:
       //       - OID of curve;
       //       - MPI of EC point representing public key.
       //       - KDF: variable-length field containing KDF parameters.
-      case enums.publicKey.ecdh:
-        return [type_oid, type_mpi, type_kdf_params];
+      case enums.publicKey.ecdh: {
+        const oid = new OID(); read += oid.read(bytes);
+        const Q = util.readMPI(bytes.subarray(read)); read += Q.length + 2;
+        const kdfParams = new KDFParams(); read += kdfParams.read(bytes.subarray(read));
+        return { read: read, params: { oid, Q, kdfParams } };
+      }
+      default:
+        throw new Error('Invalid public key encryption algorithm.');
+    }
+  },
+
+  /**
+   * Parse private key material in binary form to get the key parameters
+   * @param {module:enums.publicKey} algo  The key algorithm
+   * @param {Uint8Array}             bytes The key material to parse
+   * @param {Object}                 publicParams (ECC only) public params, needed to format some private params
```

Comment:
Maybe we should also pad RSA and DSA keys out of principle -  I don't think it's possible that they have a leading 0 byte, see [this study on RSA](https://www.usenix.org/system/files/conference/usenixsecurity16/sec16_paper_svenda.pdf)

Reply:
Yeah, the way we generate them they have at most a leading 0 bit I think, not a leading 0 byte. Regarding padding, I don't think there are many restrictions on the length of these parameters - I guess usually `p` and `q` are `n.length / 2`, but I don't think this is really required.